### PR TITLE
feat: Stack Analysis implementation in Dependency Analytics Intellij plugin

### PR DESCRIPTION
**Pull request by jparsai**
# Description
This PR is to introduce Stack Analysis capabilities of CRDA platform into existing Intellij Plugin of CRDA.

 
## Documentation
- **Approach document:-** https://docs.google.com/document/d/1dBq4bVlg_r0D7Fp_xu57JepixE2EseNP9hE-Es7fMz8/edit?usp=sharing

- **Sequence diagram for process after plugin installation/IDE restart:-** https://drive.google.com/file/d/1GVwyGnZR6KOSfYEYhhOVCJEc7lhxvBn5/view?usp=sharing

- **Sequence diagram for process to execution Stack Analysis on demand:-** https://drive.google.com/file/d/1RmFZxadK0rfbG-mbCyelAyxt_-xZaWyl/view?usp=sharing

## Type of changes in existing files

- **Readme:-** Updated details about plugin.

- **build.gradle:-** Upgraded Intellij plugin version.
Added new dependencies.

- **gradle.properties:-** Bumped up lowest supported IDE version due to JCEF loading issue in 2020.2 and 2020.3 versions of Intellij.

- **GitHubRelease.java:-** Downloading assets using name instead of label, as CLI releases are not having any label attached, downloading using name is not causing any issue in LSP download as well.

- **GitHubReleaseDownloader.java:-** Modified logic to download file based on type (LSP or CLI) instead of introducing new class to download CLI.

- **ICookie.java:-** Added new cookie for CLI.

- **Platform.java:-** Introduced new methods/variables for CLI.

- **PreloadLanguageServer.java:-** Sending repo name as an argument in GitHubReleaseDownloader() because now it is being used for CLI repo as well.

- **plugin.xml:-** Updated plugin description which is shown in Marketplace page. 
Updated build version because we need to use 2021.x + versions of IDE due to JCEF issue.
Added preloadingActivity, fileEditorProvider and set of Actions.

### CODE REVIEW

**Comment by jparsai** [comment]
Hello @jeffmaury, kindly review this PR.

File: src/main/java/org/jboss/tools/intellij/analytics/PreloadLanguageServer.java [hunk 1]
```diff
@@ -14,7 +14,7 @@
   private final ICookie cookies = ServiceManager.getService(Settings.class);
 
   private void attachLanguageClient(final File cliFile) {
-    final String[] EXTENSIONS = {"xml", "json", "txt"};
+    final String[] EXTENSIONS = {"xml", "json", "txt", "mod"};
```
Comment (jeffmaury) [inline src/main/java/org/jboss/tools/intellij/analytics/PreloadLanguageServer.java:17]: Adding Go language should not be part of this PR, you are mixing two different purposes that will render the commit log and release not difficult to read

File: src/main/java/org/jboss/tools/intellij/analytics/SaProcessExecutor.java [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+	 *
+	 * @throws IOException In case of process failure
+	 * @throws InterruptedException In case of process failure
+	 */
+	public String execute(String command) throws IOException, InterruptedException {
+		// Logic to execute given CLI command and get the result.
+		ProcessBuilder processBuilder = new ProcessBuilder();
+		// Set CLI binary location as working directory for process.
+		processBuilder.directory(new File(Platform.pluginDirectory));
+		processBuilder.command(this.getPrompt(), this.getCmd(), command);
```
Comment (jeffmaury) [inline src/main/java/org/jboss/tools/intellij/analytics/SaProcessExecutor.java:117]: Why do you need to go through the command shell to launch your process ?

File: src/main/java/org/jboss/tools/intellij/analytics/SaUtils.java [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+     *
+     * @throws IOException In case of process failure
+     */
+    public void openCustomEditor(FileEditorManager instance, String url) throws IOException {
+
+        // Close custom editor if already opened in previous run
+        closeCustomEditor(instance);
+
+        // Create a temp file in sandbox directory which is registered with SaReportEditorProvider.
+        File reportFile = new File(SaConfig.REPORT_FILE_PATH);
```
Comment (jeffmaury) [inline src/main/java/org/jboss/tools/intellij/analytics/SaUtils.java:65]: Does not seem a temp file to me. A file with a fixed file name is created in the current directory. Thus if we have projects with both pom.xml and package.json for example, this will fail. So you should create a real temp file with a specific file extension and bind the browser based editor on it.

File: src/main/java/org/jboss/tools/intellij/analytics/SaAction.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+        PsiFile psiFile = event.getData(CommonDataKeys.PSI_FILE);
+
+        // Check if file where context menu is opened is type of supported extension.
+        // If yes then show the action for SA in menu
+        // Also check Intellij version if it is lower than 2021.X then don't show SA Action.
+        if (psiFile != null) {
+            int intellijMajorVersion = Integer.parseInt(ApplicationInfo.getInstance().getMajorVersion());
+            event.getPresentation().setEnabledAndVisible(supportedManifestFiles
+                    .contains(psiFile.getName()) && intellijMajorVersion>=2021);
```
Comment (jeffmaury) [inline src/main/java/org/jboss/tools/intellij/analytics/SaAction.java:76]: The check on IntelliJ version is useless because you changed plugin minimum requirement to IJ 2021

File: src/main/java/org/jboss/tools/intellij/analytics/SaUtils.java (hunk 3 above)
Comment (jparsai) [reply src/main/java/org/jboss/tools/intellij/analytics/SaUtils.java:65]: Yes, the name of file is fixed and it is overwritten each time, the reason behind it was that temp file will be created having a random name with given prefix/suffix, only issue with it is that when file is opened in Editor, the name of file would look odd. Is it Okay with you if I create file with fixed name, but inside a unique directory for each execution? so there wont be any conflicts between execution from same project or different project. for Ex $TEMP/ProcesID or UUID/Dependency Analytics Report

PFA screenshot of Editor view in case of temp file name.

image

File: src/main/java/org/jboss/tools/intellij/analytics/SaProcessExecutor.java (hunk 2 above)
Comment (jparsai) [reply src/main/java/org/jboss/tools/intellij/analytics/SaProcessExecutor.java:117]: I am using CRDA CLI binary to run SA in background, I found that ProcessBuilder would be a good fit, there are other options as well like Runtime.exec(). If there is any better approach for it, please suggest me I will surly look into that.

Comment (jeffmaury) [reply src/main/java/org/jboss/tools/intellij/analytics/SaProcessExecutor.java:117]: Yes ProcessBuilder is ok but why don't you use it to run the CLI directly and not through the system shell (cmd.exe / sh)

File: src/main/java/org/jboss/tools/intellij/analytics/SaUtils.java (hunk 3 above)
Comment (jeffmaury) [reply src/main/java/org/jboss/tools/intellij/analytics/SaUtils.java:65]: Best way is to use a temp file and to change the editor getName() method to report correct information: ex Report for xxx (pom.xml,...)

Comment (jparsai) [reply src/main/java/org/jboss/tools/intellij/analytics/SaUtils.java:65]: I just tried to rename it using `FileEditor.getName()`, but what I found is that, it can not rename Editor, it changes the name of File tab shown in the bottom if `PLACE_AFTER_DEFAULT_EDITOR` or `PLACE_BEFORE_DEFAULT_EDITOR` `FileEditorPolicy` is used, here I am using `HIDE_DEFAULT_EDITOR` policy so file will have only one tab showing report and getName() will not be useful anymore. Please correct me if I understood it wrong.

image

Comment (jparsai) [reply src/main/java/org/jboss/tools/intellij/analytics/SaUtils.java]: As you suggested, I have changed it from a fixed file to a temp file which will be created inside temporary directory of system having a random unique name on each execution.

Comment (jparsai) [reply src/main/java/org/jboss/tools/intellij/analytics/SaUtils.java]: Able to rename the file tab using `EditorTabTitleProvider`, it will set title of temp file opened in Editor window without renaming the physical file, hence on rerun from same project/other project new temp file will be created and on opening only title will be changed  in Editor.

**Comment by jparsai** [comment]
Hello @jeffmaury, all comments are addressed PTAL.

**Review by jeffmaury — APPROVED** [review]
LGTM.
A few remarks:
- maybe worth to put code specific to the stack analysis in a separate package rather that in the root package
- would be good if the editor tab displays the original file name in the addition of "Dependency Analytics Report"
- you added a new JSON library I think IJ already provides some

**Comment by jparsai** [comment]
Sure @jeffmaury I will address your comments

`maybe worth to put code specific to the stack analysis in a separate package rather that in the root package`
Will move files related to SA in a new package

`would be good if the editor tab displays the original file name in the addition of "Dependency Analytics Report"`
Please clarify my doubt, by "original file" you mean manifest file for which SA has been triggered ("Dependency Analytics Report pom.xml") or is it the temp file you are referring ("Dependency Analytics Report CRDA-1272637366849476628.sa")?

`you added a new JSON library I think IJ already provides some`
Let me search for it, if its there I will use that instead of adding new.

**Comment by jeffmaury** [comment]
> 
> 
> Sure @jeffmaury I will address your comments
> 
> `maybe worth to put code specific to the stack analysis in a separate package rather that in the root package`
> Will move files related to SA in a new package
> 
> `would be good if the editor tab displays the original file name in the addition of "Dependency Analytics Report"`
> Please clarify my doubt, by "original file" you mean manifest file for which SA has been triggered ("Dependency Analytics Report pom.xml") or is it the temp file you are referring ("Dependency Analytics Report CRDA-1272637366849476628.sa")?
The first choice ("Dependency Analytics Report pom.xml")
> 
> `you added a new JSON library I think IJ already provides some`
> Let me search for it, if its there I will use that instead of adding new.

**Comment by yzainee** [comment]
@jeffmaury So are you ok with the approach?

**Comment by jeffmaury** [comment]
> 
> 
> @jeffmaury So are you ok with the approach?

yes, waiting for the few changes and merge would follow :)

**Comment by jparsai** [comment]
Hello @jeffmaury, I addressed your comments about Editor Tab name and individual package. Could not address the one about JSON library, I looked for it, IJ has one inbuilt but it is available in IDEA Ultimate only. PTAL

**Comment by jeffmaury** [comment]
I can find Google GSON and Jackson as part of the IC library

**Comment by jparsai** [comment]
> I can find Google GSON and Jackson as part of the IC library

Yes you are right, JSON library is there, I was searching for it at a wrong place. I have changed it. PTAL.

File: src/main/java/org/jboss/tools/intellij/stackanalysis/SaUtils.java [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+     *
+     * @param instance An instance of FileEditorManager.
+     * @param url URL to be opened in attached browser.
+     *
+     * @throws IOException In case of process failure
+     */
+    public void openCustomEditor(FileEditorManager instance, String url, String manifestName) throws IOException {
+
+        // Close custom editor if already opened in previous run
+        closeCustomEditor(instance);
```
Comment (jeffmaury) [inline src/main/java/org/jboss/tools/intellij/stackanalysis/SaUtils.java:65]: I think that needs to be removed because we should be able to display several reports since the change to dynamic temp file or we should close only the one that correspond to the same original file

Comment (jparsai) [reply src/main/java/org/jboss/tools/intellij/stackanalysis/SaUtils.java:65]: That can be done, I am following the VSCode trend, there we show new report on top of existing report. I can do this also. Yusuf WDYT?

Comment (yzainee) [reply src/main/java/org/jboss/tools/intellij/stackanalysis/SaUtils.java:65]: in vscode we always have only a single file open for the report. I think i would like to have the same behaviour here as well. multiple files will be confusing

Comment (jeffmaury) [reply src/main/java/org/jboss/tools/intellij/stackanalysis/SaUtils.java:65]: Yes but the way it is implemented today is that you can have a report for pom.xml and another to go.mod for example. So if I rerun the report for pom.xml I don't want to close the editor for go.mod

Comment (yzainee) [reply src/main/java/org/jboss/tools/intellij/stackanalysis/SaUtils.java:65]: if we can make sure that for pom.xml, only 1 tab appears and other tabs for other deps file then its ok. What i would not want is multiple tabs for pom.xml

Comment (jparsai) [reply src/main/java/org/jboss/tools/intellij/stackanalysis/SaUtils.java:65]: Okay I will change, for a particular manifest only one tab will be opened. just need one clarification, what if a project is having multiple manifests (ex <project>/requirements.txt and <project>/test/requirements.txt) or there are two project of same language in same workspace (ex <project 1>/requirements.txt and <project 2>/requirements.txt) in this case do we need to use same tab to show report or multiple?

Comment (jeffmaury) [reply src/main/java/org/jboss/tools/intellij/stackanalysis/SaUtils.java:65]: One tab per file so /a/pom.xml and /b/pom.xml should have different tabs

Comment (jparsai) [reply src/main/java/org/jboss/tools/intellij/stackanalysis/SaUtils.java]: It is done. PTAL
PFA screenshot for an example

image

**Review by jeffmaury — APPROVED** [review]
LGTM.

Are we good to merge now @jparsai ?

**Comment by jparsai** [comment]
> LGTM.
> Are we good to merge now @jparsai ?

Yes @jeffmaury, Please merge it.

**Comment by jeffmaury** [comment]
Congrats @jparsai, great work.

**Comment by jparsai** [comment]
Thank you @jeffmaury for you support and guidance.
